apis/core/lib/domain/cube-projects/import.ts

```diff
@@ -0,0 +1,121 @@
+import { cc, shape } from '@cube-creator/core/namespace'
+import $rdf from 'rdf-ext'
+import { exists } from './queries'
+import { DomainError } from '../../../../errors/domain'
+import { BadRequest } from 'http-errors'
+import { Transform } from 'stream'
+import { obj } from 'through2'
+import TermSet from '@rdfjs/term-set'
+
+export type Files = Record<string, (project: Project) => Stream>
+
+interface ImportProject {
+  projectsCollection: GraphPointer<NamedNode>
+  resource: GraphPointer
+  files: Files
+  store: ResourceStore
+  user: NamedNode
+  userName: string
+}
+
+interface ImportedProject {
+  project: Project
+  importedDataset: DatasetCore
+}
+
+const sourceKinds = new TermSet([
+  shape('cube-project/create#ExistingCube'),
+  shape('cube-project/create#CSV'),
+])
+
+/**
+ * Ensures that URI of the project itself in the stream,
+ * which should be parsed to an URI with a trailing slash,
+ * has that slash removed
+ */
+export function adjustTerms(project: Project): Transform {
+  const idWithSlash = $rdf.namedNode(project.id.value + '/')
+
+  function rewrite<T extends Term>(term: T): T {
+    if (term.equals(idWithSlash)) {
+      return project.id as any
+    }
+
+    return term
+  }
+
+  return obj(function ({ subject, predicate, object, graph }: Quad, enc, next) {
+    this.push($rdf.quad(rewrite(subject), rewrite(predicate), rewrite(object), rewrite(graph)))
+
+    next()
+  })
+}
+
+export async function importProject({
+  projectsCollection,
+  store,
+  resource,
+  files,
+  user,
+  userName,
+}: ImportProject): Promise<ImportedProject> {
+  const label = resource.out(rdfs.label).value
+  if (!label) {
+    throw new BadRequest('Missing project name')
+  }
+  const maintainer = resource.out(schema.maintainer).term
+  if (!maintainer) {
+    throw new BadRequest('Missing organization')
+  }
+
+  const projectNode = await store.createMember(projectsCollection.term, id.cubeProject(label))
+
+  const project = createMinimalProject(projectNode, {
+    creator: { id: user, name: userName },
+    maintainer,
+    label,
+  })
+
+  const importedDataset = $rdf.dataset()
+  for (const file of resource.out(cc.export).toArray()) {
+    const next = files[file.value]
+    if (!next) {
+      throw new BadRequest(`Missing file ${file.value}`)
+    }
+
+    await importedDataset.import(next(project))
+  }
+
+  const importedProject = clownface({ dataset: importedDataset, graph: project.id }).node(project.id)
+  const cubeIdentifier = importedProject.out(dcterms.identifier).value
+  if (!cubeIdentifier) {
+    throw new BadRequest('Missing cube identifier name in imported data')
+  }
+  if (await exists(cubeIdentifier, maintainer)) {
+    throw new DomainError('Another project is already using same identifier')
+  }
+  const sourceKind = importedProject.out(cc.projectSourceKind).term
+  if (!sourceKind || sourceKind.termType !== 'NamedNode' || !sourceKinds.has(sourceKind)) {
```

Comment:
I think that won't work because `sourceKinds` will have a different prefix depending on where the tool is deployed.

Reply:
Oh, you're right. It might be best to migrate the kinds to the shared `cc:` namespace ?

Reply:
Fine by me. Makes sense if they have a meaning outside of a single instance.

Reply:
They do now 😃